pyflakes/test/test_code_segment.py

```diff
@@ -124,3 +126,50 @@ def bar(f, g=1, *h, **i):
         self.assertIsInstance(function_scope_bar['g'], Argument)
         self.assertIsInstance(function_scope_bar['h'], Argument)
         self.assertIsInstance(function_scope_bar['i'], Argument)
+
+    @skipIf(version_info < (3, 5), 'new in Python 3.5')
+    def test_scope_async_function(self):
```

Comment:
fwiw the other async functions live in `test/test_other.py` -- maybe move this function there?

on second look probably doesn't need the whole test that the other function scope test has -- just enough to demonstrate that functions nested in `async def` was broken

Reply:
Hrm, I considered `test/test_other.py`, but opted for this file because the issue here was with any code when the top level element in the tree was an async function. I don't think the signature or body of the function matters; perhaps I should remove that from the test to clarify it?

Reply:
I guess I'm surprised this test didn't trigger the same bug -- so there must be something different -- https://github.com/PyCQA/pyflakes/blob/c9708a18a17fbf17e0d88c1d1675ac1a926c4565/pyflakes/test/test_other.py#L1924-L1929

also I think whatever test that demonstrates this error fits much better in the `TestAsyncStatements` test there

Reply:
Ah, I think the difference is the passing of `is_segment=True` -- that flag causes the harness to pick out the first child from the tree, meaning that the checker operates on a specific scope from the start rather than starting with a `ModuleScope` and then working its way down from there. (I initially couldn't work out why test in the commit which introduced this, and which is similar to `test_asyncDef`, didn't find the issue.)

The bug which this is fixing was to do with how `Checker.__init__` builds that root scope from the type of the initial tree node, so in most cases (which start with an `_ast.Module` and thus get a `ModuleScope`) we get away with it.

I'm guessing that this scenario would only be hit by a client using the `pyflakes` Python API directly -- certainly I couldn't construct any source which triggered it alone -- I spotted the bug from reading the code rather than experiencing the failure.

Happy to move the test and/or remove the new test and update that one if you'd prefer that, though my reasoning for the location of the test was that this issue relates to the segment behaviour as much as it does the fact that the function is async. What do you think?

Reply:
ah ok, that makes way more sense now -- then this is the correct place :+1:

Reply:
:+1: Apologies for the lack of clarify upfront.

P.S: great gifs :)

Reply:
hehe -- they come from a userscript https://github.com/chriskuehl/shipit